### PR TITLE
Expand KIM-1 samples with graphics primitives, text, etc

### PR DESCRIPTION
**Pull request by davepl**
This checkin extends the set of KIM-1 samples with a number of graphics primitives targeting the MTU K-1008 Visible Memory board.

Includes pixel, linedraw, circle, moire, and Conway's Game of Life in C.

### CODE REVIEW

File: .gitignore
Comment (mrdudz) [inline .gitignore:1]: Please remove the changes to .gitignore from this PR

Comment (mrdudz) [reply .gitignore:1]: Hello? :)

**Comment by mrdudz** [comment]
You might want to set your editor to automatically remove spaces at the end of each line...

So is this ready for merging?

**Comment by mrdudz** [comment]
hrm and there are TABs too... please fix :)

File: samples/kim1/kimHello.bin
Comment (mrdudz) [inline samples/kim1/kimHello.bin:1]: This file should be removed from the PR

File: samples/kim1/kimLife.bin
Comment (mrdudz) [inline samples/kim1/kimLife.bin:1]: This file should be removed from the PR

File: samples/kim1/kimSieve.bin
Comment (mrdudz) [inline samples/kim1/kimSieve.bin:1]: This file should be removed from the PR

File: samples/kim1/kimTest.bin
Comment (mrdudz) [inline samples/kim1/kimTest.bin:1]: This file should be removed from the PR

File: samples/kim1/ramfont.o
Comment (mrdudz) [inline samples/kim1/ramfont.o:1]: This file should be removed from the PR

File: samples/kim1/subs.o
Comment (mrdudz) [inline samples/kim1/subs.o:1]: This file should be removed from the PR

File: samples/kim1/font.rom
Comment (mrdudz) [inline samples/kim1/font.rom:1]: What is this file good for? It seems like this one should also be removed from this PR.

**Review by mrdudz — CHANGES_REQUESTED** [review]
Please remove the binaries and object files that do not belong into the repository

**Comment by mrdudz** [comment]
I fixed the codestyle issues, still those .o and .bin files should be removed. Apparently i cant remove them from the PR.

**Comment by polluks** [comment]
How about including *.lst and so on in .gitignore?

**Comment by mrdudz** [comment]
> How about including *.lst and so on in .gitignore?

I might do that later in selected directories. It's not the solution for this problem though :)

**Comment by davepl** [comment]
Thanks for the cleanup, and sorry for the mess!  Just FWIW, I had those in a local .gitignore that I was asked to remove, so did, and didn’t want to monkey with the root one.  The makefile should clean them already, and of course they should not be checked in…

- Dave

**Comment by mrdudz** [comment]
> Thanks for the cleanup, and sorry for the mess! Just FWIW, I had those in a local .gitignore that I was asked to remove, so did, and didn’t want to monkey with the root one. The makefile should clean them already, and of course they should not be checked in… - Dave

If you could just make a commit that removes those files... then i can merge :) No idea why the github interface allows me to remove textfiles no problem, but not binaries. weird.

**Comment by mrdudz** [comment]
So, what about font.rom?

**Comment by davepl** [comment]
That’s a binary of the font itself, and should remain in the project ideally.  The source contains an encoded version of it, but it comes in handy if that needs to be recreated.

- Dave

**Comment by mrdudz** [comment]
> That’s a binary of the font itself, and should remain in the project ideally. The source contains an encoded version of it, but it comes in handy if that needs to be recreated. - Dave

I see. It would be great if you could describe the samples added with this PR in samples/readme.txt in a similar way the other samples are being described.

**Comment by mrdudz** [comment]
> I see. It would be great if you could describe the samples added with this PR in samples/readme.txt in a similar way the other samples are being described.

@davepl would you? And please keep an eye on those codestyle issues. (you can always use "make checkstyle" to find those)

**Comment by mrdudz** [comment]
@davepl So?

It also might be a good idea not to keep committing into a PR once its "green" - if you can't keep it clean from breaking codestyle issues for whatever reason :)

**Comment by mrdudz** [comment]
*ping*

**Comment by mrdudz** [comment]
@davepl so what's up with this?

**Comment by davepl** [comment]
Sorry, I've been very busy and this is a little ways down the list of todos, but realize I left it unfinished.  Sorry for the inconvenience!  It all compiled and worked, but didn’t meet the project requirements, which is entirely my bad.

If you can can give me two pieces of information it would assist me in resolving it more promptly:

- What git clone command do I need to get get a current snapshot of the PR so I can revise it?
- What remains to be done other than the gitignore? I can likely do that much directly

Thanks,
Dave

**Comment by rofl0r** [comment]
the current PR is broken, thanks to the merge commits in between. never ever merge branches into a branch that's the basis for a PR. i went ahead and fixed the contents of this PR using my advanced git-fu into this single patch with is stripped of the binaries: https://0x0.st/HGj5.patch

to move forward, clone the latest master from here, create a new branch KIMList2 or whatever, and then apply above patch with git am filename.patch, then add what's been requested in another commit, and push it to github.
the outstanding request is this comment: https://github.com/cc65/cc65/pull/2194?notification_referrer_id=NT_kwDOAAU1sbE3NzQyODMzNDMzOjM0MTQyNQ#issuecomment-1765887453

@mrdudz: if you merge this PR as-is, the commit history of cc65 will be broken for eternity

**Comment by mrdudz** [comment]
> never ever merge branches into a branch that's the basis for a PR.

>  if you merge this PR as-is, the commit history of cc65 will be broken for eternity

No idea what you mean - i have been doing this all the time with the fptest branch, for example.

**Comment by mrdudz** [comment]
> the outstanding request is this comment: https://github.com/cc65/cc65/pull/2194#issuecomment-1765887453

There are also various code style issues (which is described in contributing.md)

**Comment by spiro-trikaliotis** [comment]
> > never ever merge branches into a branch that's the basis for a PR.
> 
> > if you merge this PR as-is, the commit history of cc65 will be broken for eternity
> 
> No idea what you mean - i have been doing this all the time with the fptest branch, for example.

You will get the problem once you try to merge this branch. The history will be doubled and looks very "interesting".

The better approach would be to rebase your branch to master instead of merging master into it.

But this is beyond the topic of this discussion here.

**Comment by mrdudz** [comment]
> The history will be doubled and looks very "interesting".

I see. the "broken" part confused me.

**Comment by mrdudz** [comment]
> OK, but how does one “apply the above” patch?  That’s a broader leap than I know how to make on my own, as I’m new to git.

Make a new branch, then "git am name.patch" and commit/push the changes to that branch.

> BTW, I’m content to abandon this checkin and keep the code in my own KIM-1 repo, but figured it would be useful to others in cc-65.   Perhaps it’s easier to just revert this one and avoid the headaches.  I don’t want to cause further issues.

Oh please don't, it's a nice addition to cc65

**Comment by rofl0r** [comment]
> I see. the "broken" part confused me.

you can test it on this very branch. run `git rebase -i master`, then just quit the editor that pops up with saving. git will then hickup and tell you it couldnt apply some of the duplicated commits.

> OK, but how does one “apply the above” patch?

as i said in the previous comment, using `git am filename.patch`. it will apply the patch, and create the commit from the metadata in the patch header.

**Comment by mrdudz** [comment]
> you can test it on this very branch. run `git rebase -i master`, then just quit the editor that pops up with saving. git will then hickup and tell you it couldnt apply some of the duplicated commits.

rebase seems to be something else though - all i am ever doing here is merging via the web interface. And as long as that says it can be merged - i don't see the problem.

**Comment by rofl0r** [comment]
> rebase seems to be something else though

rebase conceptually turns the commits of a branch into patches and reapplies them on wherever you point it to. if it hicks up even if you apply it on the same HEAD than the branch started on, it's quite a mess.
what happened here in particular is that OP merged his branch into itself "Merge branch 'KIMLife' of github.com:davepl/cc65 into KIMLife"

 > all i am ever doing here is merging via the web interface

i suppose you merge master into your branch. it's not that problematic, but it prevents a clean linear history, unlike rebase.

>And as long as that says it can be merged - i don't see the problem.

the problem happens for example if someone needs to git bisect a bug that appeared. with double commits like these here, the process cant complete without manual intervention.

**Comment by mrdudz** [comment]
> the problem happens for example if someone needs to git bisect a bug that appeared. with double commits like these here, the process cant complete without manual intervention.

Perhaps you should look at the history - it has not been "clean linear" for a decade now (since it is on github? i dont know) - and no problems bisecting anything so far :)

In any case, please lets stop here - this PR has other problems that need to be fixed in the first place :)

**Comment by mrdudz** [comment]
OK, since the tests passed now, it is probably ok - the most important thing for you to do now is to no more commit to this branch until it is reviewed and merged (i will have a look in the next couple days, no time to do this right now)

**Comment by mrdudz** [comment]
@davepl could you delete the romfont.o file from the PR (switch to the KIMLife branch, delete it, and commit/push) - after that i can merge this

**Comment by rofl0r** [comment]
/me prays that "squash and merge" option is used when @mrdudz clicks the merge button, else we have 60 commits like "fixe whitespace" added to history... (plus the confusing merge and double commits)

**Comment by mrdudz** [comment]
> /me prays that "squash and merge" option is used when @mrdudz clicks the merge button, else we have 60 commits like "fixe whitespace" added to history... (plus the confusing merge and double commits, and the bins will be stored in .git as well, bloating the repo size)

Not sure what was hard to understand in my previous answer to this ranting

**Comment by rofl0r** [comment]
i've offered to fix it, but apparently nobody gives a sh.. as long as "it works"

**Comment by mrdudz** [comment]
since my time is running out again.... i applied the patch locally (and removed the file), fixed the makefile (ramfont.o was missing in the clean target) and co,mitted in 934de68 - thanks!

**Pull request closed without merging**
